### ISSUE TITLE
Data standardiztion

### ISSUE THREAD

**Issue by jaelim** [Shark-ML/Shark#51]
Hi,

I hope someone with knowledge could help me on data normalization with standard distribution. Based on my first reading of the documentation, I have put together following code attempt. Obviously, it did not work. I would like to train nomalized data with random forest trainer and generate prediction with a classifier. I'm lost with proper declaration of data type.
```
    ClassificationDataset training_data, test_data;
	
	importCSV(training_data, training, LAST_COLUMN, ' ');
	importCSV(test_data, test, LAST_COLUMN, ' ');

	Normalizer<RealVector> normalizer_train, normalizer_test;
	NormalizeComponentsUnitInterval<RealVector> normalizing_train(), normalizing_test();
	normalizing_train.train(normalizer_train, training_data);
	normalizing_test.train(normalizer_test, test_data);
	
	ClassificationDataset normalizedData_train = transformInputs(training_data, normalizer_train),
					  normalizedData_test = transformInputs(test_data, normalizer_test);
```

**Comment by Ulfgard** [comment]
I do not know what "did not work" means. But what I do know that you *must* always use the same normalizer for the test data as for the training data, This is why we create the model in the first place: to make this easier.

also there is an obvious error in line 5 as you are defining a function there not a default constructed object. remove the ().

**Comment by jaelim** [comment]
Thanks for replying. Per your comment, I removed the (). Following is compilation error afterwards:
I have both training_data & test_data declared as ClassificationDataset type, but in NormalizComponentUnitInterval in the second argument place, it's in conflict with UnlabeledData<DataType>. What's proper way to use your imported data object in Normalizer such that I can obtain standardized data to be used in trainer?

`error: no matching function for call to ‘shark::NormalizeComponentsUnitInterval<shark::blas::vector<double> >::train(shark::Normalizer<shark::blas::vector<double> >&, shark::ClassificationDataset&)’
  normalizing_train.train(normalizer_train, training_data);
                                                         ^
/home/Development/data_121115/Feature_importance/rf_importance_shark.cpp:105:57: note: candidate is:
In file included from /home/Development/data_121115/Feature_importance/rf_importance_shark.cpp:2:0:
/home/Development/Shark/include/shark/Algorithms/Trainers/NormalizeComponentsUnitInterval.h:84:7: note: void shark::NormalizeComponentsUnitInterval<DataType>::train(shark::Normalizer<DataType>&, const shark::UnlabeledData<InputT>&) [with DataType = shark::blas::vector<double>]
  void train(Normalizer<DataType>& model, UnlabeledData<DataType> const& input)
       ^
/home/Development/Shark/include/shark/Algorithms/Trainers/NormalizeComponentsUnitInterval.h:84:7: note:   no known conversion for argument 2 from ‘shark::ClassificationDataset {aka shark::LabeledData<shark::blas::vector<double>, unsigned int>}’ to ‘const shark::UnlabeledData<shark::blas::vector<double> >&’
/home/Development/data_121115/Feature_importance/rf_importance_shark.cpp:106:51: error: no matching function for call to ‘shark::NormalizeComponentsUnitInterval<shark::blas::vector<double> >::train(shark::Normalizer<shark::blas::vector<double> >&, shark::ClassificationDataset&)’
  normalizing_test.train(normalizer_test, test_data);
                                                   ^
/home/Development/data_121115/Feature_importance/rf_importance_shark.cpp:106:51: note: candidate is:
In file included from /home/pilotwarela0/Development/data_121115/Feature_importance/rf_importance_shark.cpp:2:0:
/home/Development/Shark/include/shark/Algorithms/Trainers/NormalizeComponentsUnitInterval.h:84:7: note: void shark::NormalizeComponentsUnitInterval<DataType>::train(shark::Normalizer<DataType>&, const shark::UnlabeledData<InputT>&) [with DataType = shark::blas::vector<double>]
  void train(Normalizer<DataType>& model, UnlabeledData<DataType> const& input)
       ^
/home/Development/Shark/include/shark/Algorithms/Trainers/NormalizeComponentsUnitInterval.h:84:7: note:   no known conversion for argument 2 from ‘shark::ClassificationDataset {aka shark::LabeledData<shark::blas::vector<double>, unsigned int>}’ to ‘const shark::UnlabeledData<shark::blas::vector<double> >&’
make[2]: *** [CMakeFiles/rf_importance_shark.dir/rf_importance_shark.cpp.o] Error 1
make[1]: *** [CMakeFiles/rf_importance_shark.dir/all] Error 2
make: *** [all] Error 2
`

**Comment by Ulfgard** [comment]
the trainer does not care about labels, so just give it the inputs:

`normalizing_train.train(normalizer_train, training_data.inputs())'

this should fix your issue. Please remember not to train a normalizer on the test set and use

`normalizedData_test = transformInputs(test_data, normalizer_train);`

**Comment by jaelim** [comment]
@Ulfgard Hi, Thanks for replying. Could you explain why you don't want to do the same for the test set? I thought because you train your model on normalized train set, you should need to have a similar normalized test set. So, are you saying don't normalize the test set and just use it as it was for prediction?
